tensorpipe/common/cuda.h

```diff
@@ -8,14 +8,34 @@
 
 #pragma once
 
+#include <memory>
+
 #include <cuda_runtime.h>
 
 #include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+
+#define TP_CUDA_CHECK(a)                                                \
+  do {                                                                  \
+    cudaError_t error = (a);                                            \
+    TP_THROW_ASSERT_IF(cudaSuccess != error)                            \
+        << __TP_EXPAND_OPD(a) << " " << cudaGetErrorName(error) << " (" \
+        << cudaGetErrorString(error) << ")";                            \
+  } while (false)
+
+namespace tensorpipe {
+
+class CudaError final : public BaseError {
+ public:
+  CudaError(cudaError_t error) : error_(error) {}
+
+  std::string what() const override {
+    return std::string(cudaGetErrorString(error_));
+  };
 
-#define TP_CUDA_CHECK(a)                                                      \
-  TP_THROW_ASSERT_IF(cudaSuccess != (a))                                      \
-      << __TP_EXPAND_OPD(a) << " " << cudaGetErrorName(cudaPeekAtLastError()) \
-      << " (" << cudaGetErrorString(cudaPeekAtLastError()) << ")"
+ private:
+  cudaError_t error_;
+};
```

Comment:
This wraps a `cudaError_t` so that 1. we can have these set on the channel in case of failure rather than throw, and 2. we can call `CudaLoop::addCallback()` with a callback wrapped with `eagerCallbackWrapper_()`.